### PR TITLE
lreq module with tests.

### PR DESCRIPTION
**Pull request by Britefury**
First implementation of lreq module that implements primitives for representing layout space requests, along with unit tests.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+[dependencies]
+time = "0.1"
+rand = "0.3"
+
+[profile.bench]
+opt-level = 3
+debug = true
+
+# [[bin]]
```
Comment (ltratt) [inline Cargo.toml]: If this is commented out, should it be here?

File: src/lib/lspace/layout/lreq.rs [hunk 2]
```diff
@@ -0,0 +1,967 @@
...
+    if a > b {
+        return a;
+    }
+    else {
+        return b;
+    }
+}
+
+pub fn fast_maxopt<T: PartialOrd>(a: Option<T>, b: Option<T>) -> Option<T> {
+    match (a, b) {
```
Comment (ltratt) [inline src/lib/lspace/layout/lreq.rs:19]: Which is the probable common case? If it's the (Some(x), Some(y)) case, that's probably worth putting first as (I think) pattern matching is tried in order.

File: src/lib/lspace/layout/lreq.rs [hunk 3]
```diff
@@ -0,0 +1,967 @@
...
+pub fn fast_minopt<T: PartialOrd>(a: Option<T>, b: Option<T>) -> Option<T> {
+    match (a, b) {
+        (None, None) => None,
+        (x, None) => x,
+        (None, y) => y,
+        (Some(x), Some(y)) => Some(fast_min(x, y))
+    }
+}
+
+
```
Comment (ltratt) [inline src/lib/lspace/layout/lreq.rs:53]: Extraneous blank lines.

File: src/lib/lspace/layout/lreq.rs [hunk 4]
```diff
@@ -0,0 +1,967 @@
...
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum LSize {
+    Empty,
+    Size{size: f64},
+    Ref{before: f64, after: f64},
+}
+
+
```
Comment (ltratt) [inline src/lib/lspace/layout/lreq.rs:67]: Extraneous blank line.

**Comment by ltratt** [comment]
I suggest the first thing to do with this is clean up the whitespace and line length and do a force push update. I think it's worth trying "rustfmt" https://github.com/nrc/rustfmt is this may well end up being the de facto standard for all Rust code (in the same way as "go fmt").

**Comment by ltratt** [comment]
Ah, and now I notice there is such a commit. Mea culpa!

File: src/lib/lspace/layout/lreq.rs [hunk 5]
```diff
@@ -0,0 +1,957 @@
...
+        let min_size = reqs.iter().fold(None, |acc, x| fast_maxopt(acc, x.flex.min_size(&x.size)));
+        let stretch = reqs.iter().fold(None, |acc, x| fast_maxopt(acc, x.flex.stretch_opt()));
+
+        match min_size {
+            Some(m) => println!("Minimum size is {}", m),
+            _ => {}
+        };
+
+
```
Comment (ltratt) [inline src/lib/lspace/layout/lreq.rs]: Extraneous whitespace.

**Comment by ltratt** [comment]
OK, there are a couple of small comments which need addressing. 

I also wonder if some of the comments before methods could be "///" docstrings?

Then I think (possibly as a separate pull request after this one -- I don't mind) we need a travis file to run tests when pull requests are opened.

**Comment by ltratt** [comment]
Documentation commit looks good.

**Comment by Britefury** [comment]
Line lengths fixed. Come to think of it, maybe it would be best to add the new module in a separate PR...

**Comment by Britefury** [comment]
Done. Ready for review.

**Comment by ltratt** [comment]
OK, I think we should squash this down to a single commit and force push. Then it can be merged.

**Comment by Britefury** [comment]
Okay, done.

**Comment by ltratt** [comment]
Looks good.

**Pull request merged**